<!--
 * @Date: 2024-02-24 16:18:37
 * @LastEditors: zhangtiantian08 zhangtiantian08@58.com
 * @LastEditTime: 2024-02-24 17:13:52
 * @FilePath: /2023/js/async-await.html
-->
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Document</title>
    </head>
    <body>
        <script type="text/javascript">
            const promise1 = function () {
                return new Promise((resolve, reject) => {
                    setTimeout(() => {
                        return resolve(1);
                    }, 1000);
                });
            };
            const promise2 = function () {
                return new Promise((resolve, reject) => {
                    setTimeout(() => {
                        return resolve(2);
                    }, 1000);
                });
            };
            function* readFileWithGen() {
                try {
                    const content1 = yield promise1();
                    console.log("content1", content1);
                    const content2 = yield promise2();
                    console.log("content2", content2);
                    return "done";
                } catch (err) {
                    console.error(err);
                }
            }
            // var g = readFileWithGen();
            // g.next().value.then(function (data) {
            //     console.log("data :>> ", data);
            //     g.next(data).value.then(function (data) {
            //         g.next(data);
            //     });
            // });

            const run = function (generator) {
                return new Promise((resolve, reject) => {
                    const g = generator();
                    const next = function (res) {
                        const result = g.next(res);
                        if (result.done) {
                            return resolve(result.value);
                        }
                        result.value.then(
                            (res) => next(res),
                            (err) => reject(g.throw(err).value)
                        );
                    };
                    next();
                });
            };

            run(readFileWithGen)
                .then((res) => console.log(res))
                .catch((err) => console.log(err));
        </script>
    </body>
</html>
